transport/errors.go

```diff
@@ -61,44 +77,108 @@ func AsHandlerError(service, procedure string, err error) error {
 	}
 }
 
-// isHandlerError may be mixed into types which are valid HandlerErrors.
-type isHandlerError struct{}
-
-func (isHandlerError) handlerError() {}
-
 //////////////////////////////////////////////////////////////////////////////
 
 // BadRequestError is a failure to process a request because the request was
 // invalid.
-type BadRequestError struct {
-	isHandlerError
+type BadRequestError interface {
+	error
+
+	badRequestError()
+}
 
+type localBadRequestError struct {
 	Reason error
 }
 
-func (e BadRequestError) Error() string {
+var _ BadRequestError = localBadRequestError{}
+var _ HandlerError = localBadRequestError{}
+
+// LocalBadRequestError wraps the given error into a BadRequestError.
+//
+// It represents a local failure while processing an invalid request.
+func LocalBadRequestError(err error) HandlerError {
+	return localBadRequestError{Reason: err}
+}
+
+func (localBadRequestError) handlerError()    {}
+func (localBadRequestError) badRequestError() {}
+
+func (e localBadRequestError) Error() string {
 	return "BadRequest: " + e.Reason.Error()
 }
 
-// AsHandlerError on a BadRequestError returns the error as-is.
-func (e BadRequestError) AsHandlerError() HandlerError { return e }
+type remoteBadRequestError struct {
+	Message string
```

Comment:
does this need to be exported? (optionally, could this just be typed as a string)